### PR TITLE
feat: add external service

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: engine/main.go [hunk 1]
```diff
@@ -141,6 +141,7 @@ Comming soon...`,
 		conf.API.Auth.SharedInfraToken = sdk.RandomString(128)
 		conf.API.Secrets.Key = sdk.RandomString(32)
 		conf.API.Providers = append(conf.API.Providers, api.ProviderConfiguration{"sample-provider", sdk.RandomString(32)})
+		conf.API.Services = append(conf.API.Services, api.ServiceConfiguration{"sample-service", "https://ovh.github.io", 443, "/cds", 443, "doc"})
```
Comment (golangci) [inline engine/main.go]: github.com/ovh/cds/engine/api.ServiceConfiguration composite literal uses unkeyed fields

File: engine/api/api.go [hunk 2]
```diff
@@ -615,6 +626,27 @@ func (a *API) Serve(ctx context.Context) error {
...
+			HealthPath:     s.HealthPath,
+			Port:           s.Port,
+			URLWithoutPort: s.URL,
+		}
+		externalServices = append(externalServices, serv)
+	}
+	if err := services.InitExternal(a.mustDB, a.Cache, externalServices); err != nil {
+		return fmt.Errorf("unable to init external service: %v", err)
+	}
+	go services.Pings(ctx, a.mustDB, externalServices)
```
Comment (fsamin) [inline engine/api/api.go]: sdk.GoRoutine

File: engine/api/ascode_test.go [hunk 3]
```diff
@@ -58,15 +57,12 @@ func Test_postImportAsCodeHandler(t *testing.T) {
 		},
 	}))
 
-	repositoryService := services.NewRepository(func() *gorp.DbMap {
-		return db
-	}, api.Cache)
 	mockService := &sdk.Service{Name: "Test_postImportAsCodeHandler", Type: services.TypeRepositories}
-	repositoryService.Delete(mockService)
-	test.NoError(t, repositoryService.Insert(mockService))
+	services.Delete(api.mustDB(), mockService)
```
Comment (golangci) [inline engine/api/ascode_test.go]: Error return value of `services.Delete` is not checked

File: engine/api/ascode_test.go [hunk 4]
```diff
@@ -167,12 +163,9 @@ func Test_getImportAsCodeHandler(t *testing.T) {
 	api, db, _ := newTestAPI(t)
 	u, pass := assets.InsertAdminUser(db)
 
-	repositoryService := services.NewRepository(func() *gorp.DbMap {
-		return db
-	}, api.Cache)
 	mockService := &sdk.Service{Name: "Test_getImportAsCodeHandler", Type: services.TypeRepositories}
-	repositoryService.Delete(mockService)
-	test.NoError(t, repositoryService.Insert(mockService))
+	services.Delete(db, mockService)
```
Comment (golangci) [inline engine/api/ascode_test.go]: Error return value of `services.Delete` is not checked

File: engine/api/ascode_test.go [hunk 5]
```diff
@@ -233,16 +226,13 @@ func Test_postPerformImportAsCodeHandler(t *testing.T) {
 	pkey := sdk.RandomString(10)
 	_ = assets.InsertTestProject(t, db, api.Cache, pkey, pkey, u)
 
-	repositoryService := services.NewRepository(func() *gorp.DbMap {
-		return db
-	}, api.Cache)
 	mockService := &sdk.Service{Name: "Test_postPerformImportAsCodeHandler_Repo", Type: services.TypeRepositories}
-	repositoryService.Delete(mockService)
-	test.NoError(t, repositoryService.Insert(mockService))
+	services.Delete(db, mockService)
```
Comment (golangci) [inline engine/api/ascode_test.go]: Error return value of `services.Delete` is not checked

File: engine/api/services.go [hunk 6]
```diff
@@ -34,13 +36,14 @@ func (api *API) postServiceRegisterHandler() Handler {
 		}
 
 		//Insert or update the service
-		repo := services.NewRepository(api.mustDB, api.Cache)
-		if err := repo.Begin(); err != nil {
+		tx, err := api.mustDB().Begin()
+		if err != nil {
 			return sdk.WrapError(err, "postServiceRegisterHandler")
 		}
+		defer tx.Rollback()
```
Comment (golangci) [inline engine/api/services.go]: Error return value of `tx.Rollback` is not checked

File: engine/api/services.go [hunk 7]
```diff
@@ -99,16 +98,18 @@ func (api *API) serviceAPIHeartbeat(c context.Context) {
...
 
-func (api *API) serviceAPIHeartbeatUpdate(c context.Context, repo *services.Repository, hash sessionstore.SessionKey) {
-	if err := repo.Begin(); err != nil {
+func (api *API) serviceAPIHeartbeatUpdate(c context.Context, db *gorp.DbMap, hash sessionstore.SessionKey) {
+	tx, err := db.Begin()
+	if err != nil {
 		log.Error("serviceAPIHeartbeat> error on repo.Begin:%v", err)
 		return
 	}
+	defer tx.Rollback()
```
Comment (golangci) [inline engine/api/services.go]: Error return value of `tx.Rollback` is not checked

File: engine/api/ascode_test.go [hunk 8]
```diff
@@ -233,16 +226,13 @@ func Test_postPerformImportAsCodeHandler(t *testing.T) {
...
 	mockService := &sdk.Service{Name: "Test_postPerformImportAsCodeHandler_Repo", Type: services.TypeRepositories}
-	repositoryService.Delete(mockService)
-	test.NoError(t, repositoryService.Insert(mockService))
+	_ = services.Delete(db, mockService)
+	test.NoError(t, services.Insert(db, mockService))
 
 	mockService = &sdk.Service{Name: "Test_postPerformImportAsCodeHandler_VCS", Type: services.TypeHooks}
-	repositoryService.Delete(mockService)
-	test.NoError(t, repositoryService.Insert(mockService))
+	services.Delete(db, mockService)
```
Comment (golangci) [inline engine/api/ascode_test.go]: Error return value of `services.Delete` is not checked

File: engine/api/workflow/workflow_importer_test.go [hunk 9]
```diff
@@ -34,12 +32,9 @@ func TestImport(t *testing.T) {
 	key := sdk.RandomString(10)
 	proj := assets.InsertTestProject(t, db, cache, key, key, u)
 
-	repositoryService := services.NewRepository(func() *gorp.DbMap {
-		return db
-	}, cache)
 	mockService := &sdk.Service{Name: "service_test", Type: "hooks"}
-	repositoryService.Delete(mockService)
-	test.NoError(t, repositoryService.Insert(mockService))
+	services.Delete(db, mockService)
```
Comment (golangci) [inline engine/api/workflow/workflow_importer_test.go]: Error return value of `services.Delete` is not checked

**Pull request merged**